python/ray/dashboard/modules/virtual_cluster/tests/test_virtual_cluster.py

```diff
@@ -0,0 +1,457 @@
+        )
+        result = resp.json()
+        print(result)
+        return result
+    except Exception as ex:
+        logger.info(ex)
+
+
+def extract_dict_from_msg(msg: str):
+    open_brace = msg.find("{")
+    close_brace = msg.rfind("}")
+
+    if open_brace == -1 or close_brace == -1 or open_brace > close_brace:
+        raise ValueError("No valid {} pair found")
+
+    return json.loads(msg[open_brace : close_brace + 1])
+
+
+@ray.remote
+class SmallActor:
+    def pid(self):
+        return os.getpid()
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster_head",
+    [
+        {
+            "include_dashboard": True,
+        }
+    ],
+    indirect=True,
+)
+def test_create_and_update_virtual_cluster(
+    disable_aiohttp_cache, ray_start_cluster_head
+):
+    cluster: Cluster = ray_start_cluster_head
+    assert wait_until_server_available(cluster.webui_url) is True
+    webui_url = cluster.webui_url
+    webui_url = format_web_url(webui_url)
+
+    # Add two nodes to the primary cluster.
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "4c8g"})
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "8c16g"})
+    hostname = socket.gethostname()
+
+    virtual_cluster_id = ""
+    allocation_mode = ""
+    replica_sets = {}
+    revision = 0
+
+    def _create_or_update_virtual_cluster():
+        try:
+            nonlocal revision
+            resp = requests.post(
+                webui_url + "/virtual_clusters",
+                json={
+                    "virtualClusterId": virtual_cluster_id,
+                    "allocationMode": allocation_mode,
+                    "replicaSets": replica_sets,
+                    "revision": revision,
+                },
+            )
+            resp.raise_for_status()
+            result = resp.json()
+            print(result)
+            assert result["result"] is True, resp.text
+            assert result["data"]["virtualClusterId"] == virtual_cluster_id
+            current_revision = result["data"]["revision"]
+            assert current_revision > revision
+            revision = current_revision
+            virtual_cluster_replica_sets = {}
+            for _, node_instance in result["data"]["nodeInstances"].items():
+                assert node_instance["hostname"] == hostname
+                virtual_cluster_replica_sets[node_instance["templateId"]] = (
+                    virtual_cluster_replica_sets.get(node_instance["templateId"], 0) + 1
+                )
+            # The virtual cluster has the same node types and count as expected.
+            assert replica_sets == virtual_cluster_replica_sets
+            return True
+        except Exception as ex:
+            logger.info(ex)
+            return False
+
+    # Create a new virtual cluster with exclusive allocation mode.
+    virtual_cluster_id = "virtual_cluster_1"
+    allocation_mode = "exclusive"
+    replica_sets = {"4c8g": 1, "8c16g": 1}
```

Comment:
add a argument to `_create_or_update_virtual_cluster` and fill it with `replica_sets`